### ISSUE TITLE
Strategy for row-level security

### ISSUE THREAD

**Issue by IanMayo** [debrief/pepys-import#822]
The users have a requirement for the ability to store data in Pepys that can only be accessed by a small body of users.  We'll call this information `Limited Distribution`, or `LimDis`.  An extension of the concept could be that multiple bodies of information are only accessible to set groups of users.  The types of tables containing rows to be protected include all `Measurements` plus `Platforms`, `Sensors`, `Datafiles`, `Participants`, `Tasks`. `Changes` and `Logs`.  It is not expected that `References` would need to be protected.

A concrete way of doing this would be to create a separate Postgres database, with separate connection string, username, password.  But, this would prevent "power users" from doing bulk analysis across all of the data.

These seem to be potential strategies for providing this LimDis capability:

1. Separate databases, virtually joined.  This strategy would create separate databases for each LimDis block of data, but a virtual database (with separate login details) would combine the separate ones.
2. Access data via function, with credentials.  This strategy would introduce a new role that only has access to stored functions, accessed via username/password.  The function would check access permissions, and return the denormalised data.  The passwords on existing roles would be changed.  From that point on QGis/ARCGIS would only be able to connect to these functions to obtain data.
3. Row-level locking.  PostGres allows for rows to be tagged with which roles can see them.  The LimDis records would be tagged with the LimDis users for that dataset.  Ian has a worry about the performance of this, since it could require tests on  many, many `State` records.  There could be a simpler perspective.  `Datafiles` could have row-level protection.  _Maybe_ this row level protection would extend to the `State` rows that come from (refer to) that datafile.

Feedback on the above is welcome from @arunlal-v

**Comment by arunlal-v** [comment]
Let's conclude on the objective(s) to be achieved by implementing this requirement. That will give us the tools to judge/measure the strategies discussed above, or come up with new if required.

1) Users to be further classified as "Normal" and "Super" user
2) Normal users should be able to only view/edit/process information (records) that they are authorised to, irrespective of the client they use (Pepys-Importer, or PgAdmin)
3) Super users to continue using the application without any restrictions as specified in objective 2.

@IanMayo , should we rephrase or add to these objectives?

**Comment by IanMayo** [comment]
Thanks @arunlal-v  - but the model of two sets of users doesn't fit.  There could be multiple types of limited distribution.

A metaphor for this is a list of collection of files. Some may be only visible to the Finance Department, others only visible to the HR department, and others visible to all.

**Comment by arunlal-v** [comment]
@IanMayo, just a quick question.
Is **privacy_id** related to this functionality?

**Comment by IanMayo** [comment]
Yes, it _may_ relate to `Privacy`.  The issue _may_ be handled by including a specific privacy that only a subset of the users are able to view.  But, depending on your advice it may be handled through the virtual combination of two separate databases.
